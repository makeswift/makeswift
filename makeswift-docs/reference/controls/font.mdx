---
title: Font
description: The `Font` control allows you select a `fontFamily`, `fontStyle`, and `fontWeight`.
icon: "font"
---

<Frame type="glass">
  <img
    src="/images/controls/font/header.jpeg"
    alt="Font control component in Makeswift Visual Builder"
  />
</Frame>

## Params

<ParamField query="label" type="string">
  Text for the panel label in the Makeswift Visual Builder.
</ParamField>

<ParamField query="variant" type="boolean" default="true">
  A boolean to determine whether `fontStyle` and `fontWeight` are included in
  the final value. This value changes what panel inputs are shown in the
  Makeswift builder, and changes the type of defaultValue.
</ParamField>

<ParamField query="defaultValue" type="object" default="false">
  <p>The default value passed to your component when no value is available. This
  object includes three properties: </p>

<ul>
  <li>`fontFamily` - `string`</li>
  <li>`fontStyle` - `"normal" | "italic"`</li>
  <li>`fontWeight` - `number`</li>
</ul>
<p>Without `defaultValue` the data passed to your component is optional.</p>

### Example

```ts
defaultValue: {
  fontFamily: 'var(--font-grenze-gotisch)',
  fontStyle: 'italic',
  fontWeight: 700,
},
```

</ParamField>

## Prop type

The `Font` control passes an object with the following properties:

```ts
{
  fontFamily: string;
  fontStyle: string;
  fontWeight: number;
}
```
