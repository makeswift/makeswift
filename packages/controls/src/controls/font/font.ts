import { match } from 'ts-pattern'
import { z, ZodLiteral } from 'zod'

import { safeParse, type ParseResult } from '../../lib/zod'

import { ControlDataTypeKey } from '../../common'
import {
  type DeserializedRecord,
  type SerializedRecord,
} from '../../serialization'

import {
  ControlDefinition,
  serialize,
  type Resolvable,
  type SchemaType,
} from '../definition'
import { DefaultControlInstance, type SendMessage } from '../instance'

type Config = z.infer<
  | typeof Definition.schema.withVariants.relaxed.config
  | typeof Definition.schema.withVariants.strict.config
  | typeof Definition.schema.withoutVariants.relaxed.config
  | typeof Definition.schema.withoutVariants.strict.config
>

type SchemaByVariantAndDefaultValue<
  V extends Config['variant'],
  D extends Config['defaultValue'],
> = V extends false
  ? undefined extends D
    ? typeof Definition.schema.withoutVariants.relaxed
    : typeof Definition.schema.withoutVariants.strict
  : undefined extends D
    ? typeof Definition.schema.withVariants.relaxed
    : typeof Definition.schema.withVariants.strict

type Schema<C extends Config> = SchemaByVariantAndDefaultValue<
  C['variant'],
  C['defaultValue']
>
type DataType<C extends Config> = z.infer<Schema<C>['data']>
type ValueType<C extends Config> = z.infer<Schema<C>['value']>
type ResolvedValueType<C extends Config> = z.infer<Schema<C>['resolvedValue']>
type DefinitionType<C extends Config> = z.infer<Schema<C>['definition']>
type TypeType<C extends Config> = z.infer<Schema<C>['type']>

type ReturnedSchemaType<C extends Config> = {
  definition: SchemaType<DefinitionType<C>>
  type: SchemaType<TypeType<C>>
  data: SchemaType<DataType<C>>
  value: SchemaType<ValueType<C>>
  resolvedValue: SchemaType<ResolvedValueType<C>>
}

class Definition<C extends Config> extends ControlDefinition<
  typeof Definition.type,
  C,
  DataType<C>,
  ValueType<C>,
  ResolvedValueType<C>
> {
  private static readonly v1DataType = 'font::v1' as const
  private static readonly dataSignature = {
    v1: { [ControlDataTypeKey]: this.v1DataType },
  } as const
  static readonly type = 'makeswift::controls::font' as const

  static get schema() {
    const version = z.literal(1)
    const type = z.literal(this.type)

    const withVariants = z.object({
      fontFamily: z.string(),
      fontStyle: z.union([z.literal('normal'), z.literal('italic')]),
      fontWeight: z.union([
        z.literal(100),
        z.literal(200),
        z.literal(300),
        z.literal(400),
        z.literal(500),
        z.literal(600),
        z.literal(700),
        z.literal(800),
        z.literal(900),
      ]),
    })

    const withoutVariants = z.object({
      fontFamily: z.string(),
    })

    const versionedDataWithVariants = z.object({
      [ControlDataTypeKey]: z.literal(this.v1DataType),
      value: withVariants,
    })

    const versionedDataWithoutVariants = z.object({
      [ControlDataTypeKey]: z.literal(this.v1DataType),
      value: withoutVariants,
    })

    const schemas = <V, D, A extends ZodLiteral<boolean>>(
      value: SchemaType<V>,
      data: SchemaType<D>,
      variant: A,
    ) => {
      const config = z.object({
        label: z.string().optional(),
        defaultValue: value,
        variant,
      })

      const definition = z.object({
        type,
        config,
        version,
      })

      return {
        type,
        value,
        resolvedValue: value,
        data,
        config,
        version,
        definition,
      }
    }

    return {
      version,
      withoutVariants: {
        relaxed: schemas(
          withoutVariants.optional(),
          versionedDataWithoutVariants.optional(),
          z.literal<false>(false),
        ),
        strict: schemas(
          withoutVariants,
          versionedDataWithoutVariants,
          z.literal<false>(false),
        ),
      },
      withVariants: {
        relaxed: schemas(
          withVariants.optional(),
          versionedDataWithVariants.optional(),
          z.literal<true>(true),
        ),
        strict: schemas(
          withVariants,
          versionedDataWithVariants,
          z.literal<true>(true),
        ),
      },
    }
  }

  static deserialize(data: DeserializedRecord): FontDefinition {
    if (data.type !== Definition.type) {
      throw new Error(
        `Font: expected type ${Definition.type}, got ${data.type}`,
      )
    }

    const { config, version } = z
      .union([
        Definition.schema.withVariants.relaxed.definition,
        Definition.schema.withoutVariants.relaxed.definition,
      ])
      .parse(data)
    return new FontDefinition(config, version)
  }

  constructor(
    config: C,
    readonly version: z.infer<typeof Definition.schema.version>,
  ) {
    super(config)
  }

  get controlType() {
    return Definition.type
  }

  get schema(): ReturnedSchemaType<C> {
    const { withVariants, withoutVariants, ...baseSchema } = Definition.schema

    const refinedSchema = this.refinedSchema

    const schema = {
      ...baseSchema,
      type: refinedSchema.type,
      definition: refinedSchema.definition,
      value: refinedSchema.value,
      data: refinedSchema.data,
      resolvedValue: refinedSchema.resolvedValue,
    }

    return schema as ReturnedSchemaType<C>
  }

  get refinedSchema() {
    return this.config.variant === false
      ? this.config.defaultValue === undefined
        ? Definition.schema.withoutVariants.relaxed
        : Definition.schema.withoutVariants.strict
      : this.config.defaultValue === undefined
        ? Definition.schema.withVariants.relaxed
        : Definition.schema.withVariants.strict
  }

  get dataSchema(): Schema<C>['data'] {
    return this.refinedSchema.data
  }

  safeParse(data: unknown | undefined): ParseResult<DataType<C> | undefined> {
    return safeParse(this.dataSchema, data)
  }

  fromData(data: DataType<C> | undefined): ValueType<C> | undefined {
    const inputSchema = this.dataSchema.optional()
    return match(data satisfies z.infer<typeof inputSchema>)
      .with(Definition.dataSignature.v1, ({ value }) => value)
      .otherwise(() => undefined)
  }

  toData(value: ValueType<C>): DataType<C> {
    if (value == null) return undefined

    return {
      ...Definition.dataSignature.v1,
      value,
    }
  }

  copyData(data: DataType<C> | undefined): DataType<C> | undefined {
    if (data == null) return data

    return {
      ...data,
      value: {
        ...data.value,
      },
    }
  }

  resolveValue(
    data: DataType<C> | undefined,
  ): Resolvable<ResolvedValueType<C> | undefined> {
    return {
      name: Definition.type,
      readStable: () => this.fromData(data) ?? this.config.defaultValue,
      subscribe: () => () => {},
      triggerResolve: async () => {},
    }
  }

  createInstance(sendMessage: SendMessage) {
    return new DefaultControlInstance(sendMessage)
  }

  serialize(): [SerializedRecord, Transferable[]] {
    return serialize(this.config, {
      type: Definition.type,
      version: this.version,
    })
  }
}

export class FontDefinition<C extends Config = Config> extends Definition<C> {}

type DefaultValueByVariant<V extends Config['variant']> = z.infer<
  false extends V
    ? typeof Definition.schema.withoutVariants.strict.config
    : typeof Definition.schema.withVariants.strict.config
>['defaultValue']

type UserConfigRequiredDefault<V extends Config['variant']> = {
  label?: string
  variant?: V
  defaultValue: DefaultValueByVariant<V>
}

type UserConfigOptionalDefault<V extends Config['variant']> = {
  label?: string
  variant?: V
  defaultValue?: DefaultValueByVariant<V> | undefined
}

type NormedConfig<
  V extends Config['variant'],
  D extends Config['defaultValue'],
> = z.infer<SchemaByVariantAndDefaultValue<V, D>['config']>

export function Font<V extends Config['variant'] = true>(
  config: UserConfigRequiredDefault<V>,
): FontDefinition<NormedConfig<V, DefaultValueByVariant<V>>>

export function Font<V extends Config['variant'] = true>(
  config?: UserConfigOptionalDefault<V>,
): FontDefinition<NormedConfig<V, DefaultValueByVariant<V> | undefined>>

export function Font<V extends Config['variant']>(
  config?: UserConfigRequiredDefault<V> | UserConfigOptionalDefault<V>,
): FontDefinition<NormedConfig<V, DefaultValueByVariant<V> | undefined>> {
  return new FontDefinition(
    (config ? { variant: true, ...config } : { variant: true }) as NormedConfig<
      V,
      DefaultValueByVariant<V> | undefined
    >,
    1,
  )
}
