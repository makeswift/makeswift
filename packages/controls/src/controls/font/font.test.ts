import { Font, FontDefinition } from './font'

describe('Font', () => {
  describe('constructor', () => {
    test.each([
      {},
      { variant: true },
      {
        defaultValue: {
          fontFamily: 'comic sans',
          fontStyle: 'normal',
          fontWeight: 500,
        },
      },
      {
        variant: true,
        defaultValue: {
          fontFamily: 'comic sans',
          fontStyle: 'normal',
          fontWeight: 500,
        },
      },
      { variant: false },
      {
        variant: false,
        defaultValue: {
          fontFamily: 'comic sans',
        },
      },
    ])('returns correct definition', ({ variant, defaultValue }) => {
      // Arrange
      const definition = Font({ variant, defaultValue })

      // Assert
      expect(definition).toMatchSnapshot()
    })

    test.each([
      {
        fontFamily: 'comic sans',
        fontStyle: 'foo',
        fontWeight: 500,
      } as const,
      {
        fontFamily: 'comic sans',
        fontStyle: 'normal',
        fontWeight: 350,
      } as const,
    ])('non literal fontStyle and fontWeight values error', (defaultValue) => {
      // @ts-expect-error
      Font({ defaultValue })
    })

    test('disallows extraneous properties', () => {
      Font({
        label: 'Font family',
        variant: false,
        //@ts-expect-error
        foo: 'bar',
        defaultValue: {
          fontFamily: 'comic sans',
        },
      })
    })

    test('disallows extraneous sub-properties', () => {
      Font({
        label: 'Font family',
        variant: false,
        defaultValue: {
          fontFamily: 'comic sans',
          //@ts-expect-error
          foo: 'bar',
        },
      })
    })

    test('disallows extraneous sub-properties', () => {
      Font({
        label: 'Font family',
        variant: true,
        defaultValue: {
          fontFamily: 'comic sans',
          //@ts-expect-error
          foo: 'bar',
        },
      })
    })

    test('disallows extraneous sub-properties that are otherwise permissible when variant === true', () => {
      Font({
        label: 'Font family',
        variant: false,
        defaultValue: {
          fontFamily: 'comic sans',
          //@ts-expect-error
          fontStyle: 'normal',
        },
      })
    })
  })

  describe('assignability', () => {
    // Arrange
    function assignTest(_def: FontDefinition) {}

    // Assert
    assignTest(Font())
    assignTest(Font({ variant: false }))
    assignTest(Font({ variant: true }))
    assignTest(
      Font({
        variant: true,
        defaultValue: { fontFamily: '', fontStyle: 'normal', fontWeight: 400 },
      }),
    )
    assignTest(
      Font({
        variant: false,
        defaultValue: { fontFamily: '' },
      }),
    )
  })

  describe('copyData', () => {
    test('copyData when variant === true', () => {
      const data = {
        '@@makeswift/type': 'font::v1' as const,
        value: {
          fontFamily: 'comic sans',
          fontStyle: 'normal',
          fontWeight: 400,
        },
      } as const

      // Arrange
      const definition = Font()

      // Act
      const copiedData = definition.copyData(data)

      // Assert
      expect(copiedData).not.toBe(data)
      expect(copiedData).toEqual(data)
    })

    test('copyData when variant === false', () => {
      const data = {
        '@@makeswift/type': 'font::v1' as const,
        value: {
          fontFamily: 'comic sans',
        },
      }

      // Arrange
      const definition = Font({ variant: false })

      // Act
      //@ts-expect-error
      Font().copyData(data)

      const copiedData = definition.copyData(data)

      // Assert
      expect(copiedData).not.toBe(data)
      expect(copiedData).toEqual(data)
    })
  })

  describe('fromData', () => {
    test.each([
      {
        config: { variant: true },
        rawData: {
          '@@makeswift/type': 'font::v1' as const,
          value: {
            fontFamily: 'comic sans',
            fontStyle: 'normal',
            fontWeight: 400,
          },
        },
      },
      {
        config: {},
        rawData: {
          '@@makeswift/type': 'font::v1' as const,
          value: {
            fontFamily: 'comic sans',
            fontStyle: 'normal',
            fontWeight: 400,
          },
        },
      },
      {
        config: { variant: false },
        rawData: {
          '@@makeswift/type': 'font::v1' as const,
          value: {
            fontFamily: 'comic sans',
          },
        },
      },
    ] as const)('`%s`', ({ config, rawData }) => {
      const definition = Font(config)
      // Arrange
      const val = definition.fromData(rawData)

      // Act
      const data = definition.toData(val)

      // Assert
      expect(data).toEqual(rawData)
    })

    test.each([
      { variant: true },
      { variant: false },
      {
        variant: true,
        defaultValue: {
          fontFamily: 'comic sans',
          fontStyle: 'normal',
          fontWeight: 400,
        },
      },
      { variant: false, defaultValue: { fontFamily: 'comic sans' } },
    ] as const)('gracefully handle undefined `%s`', (config) => {
      const definition = Font(config)
      // Arrange
      const val = definition.fromData(undefined)

      // Act
      const data = definition.toData(val)

      // Assert
      expect(data).toEqual(undefined)
    })
  })

  describe('toData', () => {
    test.each([
      {
        config: { variant: true },
        value: {
          fontFamily: 'comic sans',
          fontStyle: 'normal',
          fontWeight: 400,
        },
      },
      {
        config: {},
        value: {
          fontFamily: 'comic sans',
          fontStyle: 'normal',
          fontWeight: 400,
        },
      },
      {
        config: { variant: false },
        value: {
          fontFamily: 'comic sans',
        },
      },
    ] as const)('`%s`', ({ config, value }) => {
      // Arrange
      const definition = Font(config)

      // Assert
      expect(definition.toData(value)).toMatchSnapshot()
    })

    test.each([
      { variant: true },
      { variant: false },
      {
        variant: true,
        defaultValue: {
          fontFamily: 'comic sans',
          fontStyle: 'normal',
          fontWeight: 400,
        },
      },
      { variant: false, defaultValue: { fontFamily: 'comic sans' } },
    ] as const)('gracefully handle undefined `%s`', (config) => {
      // Arrange
      const definition = Font(config)

      // Assert
      expect(definition.toData(undefined)).toMatchSnapshot()
    })
  })
})
