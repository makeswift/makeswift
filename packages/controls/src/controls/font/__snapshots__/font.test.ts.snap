// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Font constructor returns correct definition 1`] = `
FontDefinition {
  "config": {
    "defaultValue": undefined,
    "variant": undefined,
  },
  "version": 1,
}
`;

exports[`Font constructor returns correct definition 2`] = `
FontDefinition {
  "config": {
    "defaultValue": undefined,
    "variant": true,
  },
  "version": 1,
}
`;

exports[`Font constructor returns correct definition 3`] = `
FontDefinition {
  "config": {
    "defaultValue": {
      "fontFamily": "comic sans",
      "fontStyle": "normal",
      "fontWeight": 500,
    },
    "variant": undefined,
  },
  "version": 1,
}
`;

exports[`Font constructor returns correct definition 4`] = `
FontDefinition {
  "config": {
    "defaultValue": {
      "fontFamily": "comic sans",
      "fontStyle": "normal",
      "fontWeight": 500,
    },
    "variant": true,
  },
  "version": 1,
}
`;

exports[`Font constructor returns correct definition 5`] = `
FontDefinition {
  "config": {
    "defaultValue": undefined,
    "variant": false,
  },
  "version": 1,
}
`;

exports[`Font constructor returns correct definition 6`] = `
FontDefinition {
  "config": {
    "defaultValue": {
      "fontFamily": "comic sans",
    },
    "variant": false,
  },
  "version": 1,
}
`;

exports[`Font toData \`{ config: [Object], value: [Object] }\` 1`] = `
{
  "@@makeswift/type": "font::v1",
  "value": {
    "fontFamily": "comic sans",
    "fontStyle": "normal",
    "fontWeight": 400,
  },
}
`;

exports[`Font toData \`{ config: [Object], value: [Object] }\` 2`] = `
{
  "@@makeswift/type": "font::v1",
  "value": {
    "fontFamily": "comic sans",
  },
}
`;

exports[`Font toData \`{ config: {}, value: [Object] }\` 1`] = `
{
  "@@makeswift/type": "font::v1",
  "value": {
    "fontFamily": "comic sans",
    "fontStyle": "normal",
    "fontWeight": 400,
  },
}
`;

exports[`Font toData gracefully handle undefined \`{ variant: false }\` 1`] = `undefined`;

exports[`Font toData gracefully handle undefined \`{ variant: false, defaultValue: [Object] }\` 1`] = `undefined`;

exports[`Font toData gracefully handle undefined \`{ variant: true }\` 1`] = `undefined`;

exports[`Font toData gracefully handle undefined \`{ variant: true, defaultValue: [Object] }\` 1`] = `undefined`;
